lib/ansible/modules/identity/cyberark/cyberark_authentication.py

```diff
@@ -0,0 +1,301 @@
+
+    except Exception as e:
+
+        module.fail_json(
+            msg="end_point=" +
+            api_base_url +
+            end_point,
+            payload=payload,
+            headers=headers,
+            exception=to_text(e),
+            status_code=-1)
+
+    if response.getcode() == 200:  # Success
+
+        if state == "present":  # Logon Action
+
+            # Result token from REST Api uses a different key based
+            # the use of shared logon authentication
+            if use_shared_logon_authentication:
+                token = json.loads(response.read())["LogonResult"]
+            else:
+                token = json.loads(response.read())["CyberArkLogonResult"]
+
+            # Preparing result of the module
+            result = {
+                "cyberark_session": {
+                    "token": token,
+                    "api_base_url": api_base_url,
+                    "validate_certs": validate_certs,
+                    "use_shared_logon_authentication":
+                    use_shared_logon_authentication},
+            }
+
+            if new_password is not None:
+                # Only marks change if new_password was received resulting
+                # in a password change
+                changed = True
+
+        else:  # Logoff Action clears cyberark_session
+
+            result = {
+                "cyberark_session": {}
+            }
+
+        return (changed, result, response.getcode())
+
+    else:
+        module.fail_json(
+            msg="error in end_point=>" +
+            end_point,
+            headers=headers)
+
+
+def main():
+
+    fields = {
+        "api_base_url": {"type": "str"},
+        "validate_certs": {"type": "bool",
+                           "default": "true"},
+        "username": {"type": "str"},
+        "password": {"type": "str", "no_log": True},
+        "new_password": {"type": "str", "no_log": True},
+        "use_shared_logon_authentication": {"default": False, "type": "bool"},
+        "use_radius_authentication": {"default": False, "type": "bool"},
+        "state": {"type": "str",
+                  "choices": ["present", "absent"],
+                  "default": "present"},
+        "cyberark_session": {"type": "dict"},
+    }
+
+    mutually_exclusive = [
+        ["use_shared_logon_authentication", "use_radius_authentication"],
+        ["use_shared_logon_authentication", "new_password"],
+        ["api_base_url", "cyberark_session"],
+        ["cyberark_session", "username", "use_shared_logon_authentication"]
+    ]
+
+    required_if = [
+        ("state", "present", ["api_base_url"]),
+        ("state", "absent", ["cyberark_session"])
+    ]
+
+    required_together = [
+        ["username", "password"]
+    ]
+
+    module = AnsibleModule(
+        argument_spec=fields,
+        mutually_exclusive=mutually_exclusive,
+        required_if=required_if,
+        required_together=required_together,
+        supports_check_mode=True)
```

Comment:
I like the code organization here 👍 , but it is a little different than most modules. Some folks may prefer the deeply nested args style. I don't think you should change it, just a heads up that it might be a little unusual to some folks.